### PR TITLE
sys-apps/less: add 643-r1 (add certtool filters and system filter support)

### PR DESCRIPTION
**Pull request by vpayno**
1. Use certtool instead of openssl when appropriate

   - openssl only shows you the first certificate/entry it encounters
     and ignores the rest. Using certtool will correctly display all
     the entries in a single file. The view is also more readable with
     certtool.

   - Since openssl is still used as a fallback, not forcing a runtime
     dependency on `net-libs/gnutls[tools]`.

   - Java keystore filters were missing, so add them (jks/legacy and                                                                                                    
     pfx/p12).                                                                                                                                                              
                                                                                      
   - Also added a p12 file filter.     
   
2. Add support for system filters

   - Looking for a way to add custom, system wide, filters to an image
     from other ebuilds (so the lessfilter is installed only when that
     package is installed).

@chutz 
@williamh 

```diff
$ diff -uNr files/lesspipe-r{2,3}.sh                                                                                                                               [71/4581]
--- files/lesspipe-r2.sh        2023-11-28 18:44:39.958640336 -0800
+++ files/lesspipe-r3.sh        2023-11-28 20:48:52.933575496 -0800
@@ -24,6 +24,42 @@                                                                                                                                                          
        esac     
 }                                   
  
+crl_filter() {                                                                                                                                                             +    if command -v certtool &>/dev/null; then                                                                                                                               +        certtool --crl-info --text --infile "$1"                                                                                                                           +    else                   
+        openssl crl -hash -text -noout -in "$1"
+    fi            
+}
+                                                                                     
+csr_filter() {
+    if command -v certtool &>/dev/null; then
+        certtool --crq-info --text --infile "$1"    
+    else                                                                             
+        openssl req -text -noout -in "$1"                                            
+    fi                                                                               
+}                                                                                    
+                                                                                     
+crt_filter() {                                                                       
+    if command -v certtool &>/dev/null; then                
+        certtool --certificate-info --text --infile "$1"
+    else
+        openssl x509 -hash -text -noout -in "$1"                                   
+    fi                                                                               
+}
+                                                                                                                                                                           
+jks_filter() {  
+    if command -v keytool &>/dev/null; then
+        keytool -list -keystore "$1"
+    else                                                                                                                                                                   +        cat "$1"                                                                                                                                                           +    fi                                                                                                                                                                     +}                          
+
+p12_filter() {    
+    openssl pkcs12 -nokeys -info -in "$1"
+}                                                                                    
+                                                                                     
 lesspipe_file() {                                                                    
        local out=$(file -L -- "$1")                                                  
        local suffix
@@ -57,6 +93,15 @@
                ~/.lessfilter "$1" && exit 0
        fi
                                                                                      
+       # System filters         
+       if compgen -G "/etc/lessfilter.d/*.sh"; then
+               for f in /etc/lessfilter.d/*.sh; do 
+                       if [[ -x ${f} ]]; then
+                               "${f}" "$1" && exit 0
+                       fi
+               done
+       fi                                                                            
+                                                                                     
        local ignore
        for ignore in ${LESSIGNORE} ; do                                                                                                                                    
                [[ ${match} == *.${ignore} ]] && exit 0
@@ -199,9 +244,11 @@                 
                ;;
                                                                                                                                                                                    ### Encryption stuff ###                                                                                                                                            -       *.crl) openssl crl -hash -text -noout -in "$1" ;;                                                                                                                   -       *.csr) openssl req -text -noout -in "$1" ;;
-       *.crt|*.pem) openssl x509 -hash -text -noout -in "$1" ;;
+       *.crl) crl_filter "$1" ;;
+       *.csr) csr_filter "$1" ;;
+       *.crt|*.pem) crt_filter "$1" ;;                                             
+       *.jks) jks_filter "$1" ;;
+       *.p12|*.pfx) p12_filter "$1" ;;
  
 # May not be such a good idea :)
 #      ### Device nodes ###
```

```diff
$ diff -uNr less-643.ebuild less-643-r1.ebuild
--- less-643.ebuild     2023-11-28 18:44:39.958640336 -0800
+++ less-643-r1.ebuild  2023-11-28 20:22:17.324490381 -0800
@@ -75,7 +75,9 @@
 src_install() {
        default
 
-       newbin "${FILESDIR}"/lesspipe-r2.sh lesspipe
+       keepdir /etc/lessfilter.d
+
+       newbin "${FILESDIR}"/lesspipe-r3.sh lesspipe
        newenvd "${FILESDIR}"/less.envd 70less
 }
```

### CODE REVIEW

**Comment by thesamesam** [comment]
Please include some rationale in the commit message.

**Comment by vpayno** [comment]
> Please include some rationale in the commit message.

@thesamesam done

File: sys-apps/less/files/lesspipe-r3.sh [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+	local DECOMPRESSOR=$(guesscompress "${match}")
+
+	# User filters
+	if [[ -x ~/.lessfilter ]] ; then
+		~/.lessfilter "$1" && exit 0
+	fi
+
+	# System filters
+	if compgen -G "/etc/lessfilter.d/*.sh"; then
```
Comment (Flowdalic) [inline sys-apps/less/files/lesspipe-r3.sh:97]: `compgen` requires `RDEPEND=app-shells/bash[readline]`, see https://bugs.gentoo.org/909163. But maybe we can avoid using `compgen` here?

Comment (vpayno) [reply sys-apps/less/files/lesspipe-r3.sh:97]: I would suggest removing the `readline` use flag from the `gentoo/stage3` container to make this "issue" more obvious.

Using nullglob instead, not testing/refactoring the whole script so it's localized to just this change.

```diff
$ git show
commit 5e16f0f7562bc7405fc08bea4b3c803bdfb709cb (HEAD -> sys-apps_less_lesspipe)
Author: Victor Payno <vpayno@users.noreply.github.com>
Date:   Fri Dec 1 12:50:34 2023 -0800

    fixup! sys-apps/less: add 643-r1 (add certtool filters and system filter support)

diff --git a/sys-apps/less/files/lesspipe-r3.sh b/sys-apps/less/files/lesspipe-r3.sh
index 0bffa3313e0a..24166064425d 100644
--- a/sys-apps/less/files/lesspipe-r3.sh
+++ b/sys-apps/less/files/lesspipe-r3.sh
@@ -94,13 +94,13 @@ lesspipe() {
        fi
 
        # System filters
-       if compgen -G "/etc/lessfilter.d/*.sh"; then
-               for f in /etc/lessfilter.d/*.sh; do
-                       if [[ -x ${f} ]]; then
-                               "${f}" "$1" && exit 0
-                       fi
-               done
-       fi
+       shopt -s nullglob
+       for f in /usr/lib/lessfilter.d/*; do
+               if [[ -x ${f} ]]; then
+                       "${f}" "$1" && exit 0
+               fi
+       done
+       shopt -u nullglob
 
        local ignore
        for ignore in ${LESSIGNORE} ; do
```

[the remark this answers is not in the document]
Comment (vpayno) [reply sys-apps/less/files/lesspipe-r3.sh:97]: Sure, the filter scripts/binaries aren't configuration files so they should be installed under `/usr`.

```diff
$ git show
commit 5e16f0f7562bc7405fc08bea4b3c803bdfb709cb (HEAD -> sys-apps_less_lesspipe)
Author: Victor Payno <vpayno@users.noreply.github.com>
Date:   Fri Dec 1 12:50:34 2023 -0800

    fixup! sys-apps/less: add 643-r1 (add certtool filters and system filter support)

diff --git a/sys-apps/less/files/lesspipe-r3.sh b/sys-apps/less/files/lesspipe-r3.sh
index 0bffa3313e0a..24166064425d 100644
--- a/sys-apps/less/files/lesspipe-r3.sh
+++ b/sys-apps/less/files/lesspipe-r3.sh
@@ -94,13 +94,13 @@ lesspipe() {
        fi
 
        # System filters
-       if compgen -G "/etc/lessfilter.d/*.sh"; then
-               for f in /etc/lessfilter.d/*.sh; do
-                       if [[ -x ${f} ]]; then
-                               "${f}" "$1" && exit 0
-                       fi
-               done
-       fi
+       shopt -s nullglob
+       for f in /usr/lib/lessfilter.d/*; do
+               if [[ -x ${f} ]]; then
+                       "${f}" "$1" && exit 0
+               fi
+       done
+       shopt -u nullglob
 
        local ignore
        for ignore in ${LESSIGNORE} ; do
```

[the remark this answers is not in the document]
Comment (chutz) [reply sys-apps/less/files/lesspipe-r3.sh]: Maybe do both `/usr/lib/lessfilter.d/*` and `/etc/lessfilter.d/*`

[the remark this answers is not in the document]
Comment (vpayno) [reply sys-apps/less/files/lesspipe-r3.sh]: Just noticed I forgot to update the ebuild.

Sure, readding `/etc/lessfilter.d/`.

```diff
commit a034c75b63def6799f965dc77c46f21e61a18adc (HEAD -> sys-apps_less_lesspipe)
Author: Victor Payno <vpayno@users.noreply.github.com>
Date:   Fri Dec 1 15:52:06 2023 -0800

    fixup! sys-apps/less: add 643-r1 (add certtool filters and system filter support)

diff --git a/sys-apps/less/files/lesspipe-r3.sh b/sys-apps/less/files/lesspipe-r3.sh
index 24166064425d..33316c14a0f3 100644
--- a/sys-apps/less/files/lesspipe-r3.sh
+++ b/sys-apps/less/files/lesspipe-r3.sh
@@ -95,7 +95,7 @@ lesspipe() {
 
        # System filters
        shopt -s nullglob
-       for f in /usr/lib/lessfilter.d/*; do
+       for f in /usr/lib/lessfilter.d/* /etc/lessfilter.d/*;  do
                if [[ -x ${f} ]]; then
                        "${f}" "$1" && exit 0
                fi
diff --git a/sys-apps/less/less-643-r1.ebuild b/sys-apps/less/less-643-r1.ebuild
index e90eddb8c99a..186eda3b6814 100644
--- a/sys-apps/less/less-643-r1.ebuild
+++ b/sys-apps/less/less-643-r1.ebuild
@@ -75,6 +75,7 @@ src_test() {
 src_install() {
        default
 
+       keepdir /usr/lib/lessfilter.d
        keepdir /etc/lessfilter.d
 
        newbin "${FILESDIR}"/lesspipe-r3.sh lesspipe
```

**Comment by thesamesam** [comment]
thanks!

**Pull request closed without merging**